### ISSUE TITLE
[UX/Windows] Use last stable MSVC/CL by default + shortcut name for peview and other versions

### ISSUE THREAD

**Issue by Klaim** [build2/build2#130]
Currently `build2` will look for the most recent `cl` version installed on the system when `cl` is specified (or set by default).
This is excellent behavior, very useful, except when working on projects which are not experimentations and that require to be built with stable versions of MSVC/cl while there is a preview of MSVC/cl installed. In that situation, `build2` will always prefer the preview version (which is not -or less- stable the ... stable version), probably because indeed it is a higher version.

Still, it's super useful to have quick access to the preview version (or any other version installed).

Propositions:
---------------

A. Unless specified by the user, always use the last stable version of `cl` by default.

"Normal" work on projects usually require using a stable version of MSVC/cl.

B. Provide a name for the preview versions of MSVC/cl

Something like `cl-preview` for example, would select the last preview version installed.
That would make working using the preview still almost as easy as currently: `bdep init -C build-msvc-preview cc config.cxx=cl-preview`.

I'm not at ease exactly with a name specific to that compiler, maybe some generalization of the idea would help, though on linux the different compilers usually have different names, so the issue doesn't appear at all...

C. Provide a way to specify a specific version (familly) of MSVC/cl:

I'm thinking about something like `cl-19` for the [current versions](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B) (which can change at each update of msvc).
This is mainly to help with working with [various MSVC versions](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B), which is required for some of the projects I have worked on in the past.
Example: `bdep init -C build-vs2017 cc config.cxx=cl-14 ` 

and/or "bonus" version: through `msvc` aliases that know which cl version to use: `bdep init -C build-vs2017 cc config.cxx=msvc-2017` translated to `cl-14`

-------

It is currently possible to just specify the compiler path, so take these propositions mainly as convenient shortcuts for MSVC users to work faster, in particular when setting up configurations for a project.

**Comment by kamrann** [comment]
I agree that default behaviour when just specifying `config.cxx = cl` should use latest stable.

Further specification has considerable complexities related to bizarre MSVC versioning, and there are various approaches that could be taken. Microsoft has separate (but linked) version numbering for its IDE and its compiler toolchain. Then there is the further complexity of product variants (Community, Professional, Preview). My understanding though, is that from `build2`'s point of view, the only thing that matters is the toolchain version (what follows is based on this assumption, so let me know if you think it's incorrect).

VS will install to something like `.../Microsoft Visual Studio/[IDE version, eg. '2022']/[Variant, eg. 'Preview']`. Then below that there will be `/VC/Tools/MSVC/`, and in that directory a folder named for each installed toolchain version, such as `14.32.31326`. Some toolchain versions may appear in multiple places, but I do not believe `build2` need care about whether a toolchain is under `2022` vs `2019`, or `Community` vs `Preview`.

While some build systems (eg. [CMake](https://cmake.org/cmake/help/latest/variable/MSVC_TOOLSET_VERSION.html)) expose MSVC toolchain indirectly via IDE version (or anyway a toolset version family tied to the IDE version), I've found this both limiting and confusing. My suggestion would be to have essentially two modes:
1. Default, latest stable, as suggested. Specified by just `config.cxx = cl`.
2. Specified toolchain, in the form of:
```
config.cxx = cl
config.cxx.toolchain = 14.32.31326
```
We could also support, for example, `config.cxx.toolchain = 14.32` to mean latest installed subversion of `14.32`.

I suspect most people using a Preview installation, or wanting anything other than default stable toolchain, will probably be aware of the toolchain locations and able to set this config value. That said, we could also have special values for `cxx.toolchain` to make it easier; `latest`, and `stable` (the implicit default).

**Comment by boris-kolpackov** [comment]
> Some toolchain versions may appear in multiple places, but I do not believe `build2` need care about whether a toolchain is under 2022 vs 2019, or Community vs Preview.

I've seen Microsoft change the compiler version (as printed by `cl.exe` itself) without changing the "directory version" and when I asked Microsoft folks if this is a bug, they said it is "by design". So I don't think this holds.

> My suggestion would be to have essentially two modes:

We already have two modes: just `cl` or absolute path. So this will probably be a third mode.

> Specified toolchain, in the form of:
>
>     config.cxx = cl
>     config.cxx.toolchain = 14.32.31326

I think this is a sensible semantics but the syntax feels too special-case'y to me: while `config.cxx.toolchain` looks like a generic mechanism while being MSVC-specific. I think a better approach would be to work the toolchain version into the `config.{c,cxx}` value itself. In fact, something like `config.cxx = cl-14.32.31326` would be pretty consistent with how the rest of the world works (e.g., `g++-12`, `clang++-14`).

Also a note on terminology: MSVC seem to call this "build tools" (at least in the installer) while CMake -- "toolset".

**Comment by kamrann** [comment]
> I've seen Microsoft change the compiler version (as printed by `cl.exe` itself) without changing the "directory version" and when I asked Microsoft folks if this is a bug, they said it is "by design". So I don't think this holds.

Indeed, I've run into this and wasted a bunch of time dealing with it and its interaction with Unreal Engine's build tool. My impression is that they will do this for very minor build number bumps, and that it makes more sense to use the "directory version" for identification purposes rather than the precise version of the binary.

> We already have two modes: just cl or absolute path. So this will probably be a third mode.

Yep, I guess I was looking at it from the point of view that specifying toolchain version was more explicit and would remove need for the 'full path to executable' mode. Just in the sense that it feels a little strange to me to provide the `cl.exe` path, when what you're actually doing is defining the toolchain version (`cl` binary but also associated `lib`/`include` folders). On the other hand, I do accept that if the same version of the toolchain existed under both `/2019` and `/2022` installations, for example, then it becomes a little strange if it's chance as to which one gets used, even assuming they are identical.

> I think this is a sensible semantics but the syntax feels too special-case'y to me: `config.cxx.toolchain` looks like a generic mechanism while being MSVC-specific. I think a better approach would be to work the toolchain version into the `config.{c,cxx}` value itself. In fact, something like `config.cxx = cl-14.32.31326` would be pretty consistent with how the rest of the world works (e.g., `g++-12`, `clang++-14`).

Okay, yeah if other compilers have no equivalent and that's the standard way to identify them, then yes that approach seems more consistent.

**Comment by boris-kolpackov** [comment]
> On the other hand, I do accept that if the same version of the toolchain existed under both `/2019` and `/2022` installations, for example, then it becomes a little strange if it's chance as to which one gets used, even assuming they are identical.

Feels like picking it from the latest version of MSVC would be sensible in this case (with absolute `cl.exe` path serving as an escape hatch).

> you're actually doing is defining the toolchain version (`cl` binary but also associated `lib`/`include` folders).

Speaking of which, there is also Platform SDK (aka Windows SDK) of which one can have multiple versions. In your experience, is this something that could be desirable to override?

**Comment by Klaim** [comment]
> Speaking of which, there is also Platform SDK (aka Windows SDK) of which one can have multiple versions. In your experience, is this something that could be desirable to override?

Yes. It's useful when you rely on specific features existing, say, in Windows 11 but not Windows 10.

Although personally I don't have this kind of requirement every year. Depends a lot on the project. Note that since Windows 8 the DirectX SDK is part of the Windows SDK so having control on which version of the SDK to use is important to gamedevs and people working on graphics (DirectX12 being the Windows equivalent of Vulkan, to draw a broad picture of the situation).

**Comment by Klaim** [comment]
> Speaking of which, there is also Platform SDK (aka Windows SDK) of which one can have multiple versions. In your experience, is this something that could be desirable to override?

Yes, it's useful to be able to constrain which SDK to use when you rely on APIs which are for example available in Windows 11 but not in Windows 10 (or even minor versions of these). Although I rarely had to do that specifically.

More importantly, DirectX is part of the Windows SDK since Windows 8, which means graphic engine developers (and basically all gamedevs using C++) need to specify which Windows SDK to use to obtain the right DirectX 12 version (which is the equivalent of Vulkan but specific to Microsoft platforms, to draw a broad picture of the situation).

So I think it's worth it if possible, just lesser priority than being able to select the toolchain/toolset/build-tools.

> I think a better approach would be to work the toolchain version into the config.{c,cxx} value itself. In fact, something like config.cxx = cl-14.32.31326 would be pretty consistent with how the rest of the world works (e.g., g++-12, clang++-14).

I agree too with this, it's basically what I proposed in section C of the initial post, except I would like to be able to not have to specify the patch versions as usually they are very compatible (not the true for major.minor part).
Also note that in section B I proposed to have a specific way to get the preview version because otherwise you can have twice the same version, one in preview, one in community, with the same number, the "preview" part being the only difference. Also selecting "just take the last preview version on the system" is nice, the same way as "just take the last stable version on the system".

Another note: in most of my projects I need to update visual studio versions as soon as they are available, so I still need ways to make my projects use either stable of preview, whatever the version. I'm mentioning this because the directory paths changes very often these days as they push new versions at least every months (maybe more?) but I don't really care personally, it's never breaking my projects (it was also the way I was maintaining some other older projects. Both being able to stay on a specific version and being able to upgrade aggressively are important, in my opinion. This matches a lot of the rest of the discussion so I just wanted to make sure we agree on some of these points at least.

**Comment by kamrann** [comment]
> Feels like picking it from the latest version of MSVC would be sensible in this case (with absolute `cl.exe` path serving as an escape hatch).

Yep, but as @Klaim noted, there is also Community vs Professional vs Preview, where the default choice is less obvious.

> Yes, it's useful to be able to constrain which SDK to use when you rely on APIs which are for example available in Windows 11 but not in Windows 10 (or even minor versions of these). Although I rarely had to do that specifically.

Same, don't recall having had a need myself, but makes sense to be able to do it.
 
> Also note that in section B I proposed to have a specific way to get the preview version because otherwise you can have twice the same version, one in preview, one in community, with the same number, the "preview" part being the only difference. 

Yep but I think the question to clarify here is, does it matter? I believe if you have the same version installed in both, then they will be identical so being able to specify "take version X from preview installation" would be redundant. From my understanding, the difference is just that in a preview installation you will be able to install the most recent versions, which won't yet be available in a non-preview installation. 
Definitely agree that "latest preview" should be an option though.

> Both being able to stay on a specific version and being able to upgrade aggressively are important, in my opinion. This matches a lot of the rest of the discussion so I just wanted to make sure we agree on some of these points at least.

Agreed.

**Comment by Klaim** [comment]
> Yep but I think the question to clarify here is, does it matter? I believe if you have the same version installed in both, then they will be identical so being able to specify "take version X from preview installation" would be redundant. From my understanding, the difference is just that in a preview installation you will be able to install the most recent versions, which won't yet be available in a non-preview installation.

Preview, Community and Professional versions don't output the same name when looking at versions, mostly because of their respective "channel" being part of their name. While it have no consequences for direct users (because yes the binaries should be almost the same, just different internal strings with the channel name in them), I suspect it might lead to issues for other tools trying to figure out what version is being used?

I'm not sure, it's just a supposition. Maybe it's not important. Depends how these names are used outside `build2` I guess.

**Comment by Klaim** [comment]
To be clear, I'm not against assuming any same-version from whatever channel can be picked. I suspect there might be issues, but it can be dealt with later and it's easier to assume they are the same, I'm ok with that. We can be more "picky" later I think.

**Comment by kamrann** [comment]
> Preview, Community and Professional versions don't output the same name when looking at versions, mostly because of their respective "channel" being part of their name. 

Okay, then it's even more of a convoluted mess than I realized, and I'm not sure whether it would be best to require specifying channel or not at this point.

[This MSVC article](https://devblogs.microsoft.com/cppblog/side-by-side-minor-version-msvc-toolsets-in-visual-studio-2017/#using-a-side-by-side-minor-version-msvc-toolset-from-the-command-line) is a little old but may be of use. In particular it details how you can specify tools version for the VC command prompt, which might possibly be useful for implementation.

**Comment by Klaim** [comment]
Wait, correction: `cl` alone does give the same version without stating `preview` or `community` , it's the cmd/powershell envs that don't, on startup prompt.
Sorry, I mixed up.
This makes it less a problem I think?

**Comment by boris-kolpackov** [comment]
I am pretty sure things under `.../VC/Tools/MSVC/` are exactly the same regardless of Community/Professional/etc.

**Comment by boris-kolpackov** [comment]
Another data point: the MSVC Installer for 17.3.5 now offers "Windows 11 SDK" in addition to "Windows 10 SDK". The version of 11 is `10.0.22621.0` while 10 -- `10.0.19041.0`. Looking at other available versions, it feels like anything from `10.0.2200.0` is 11. Versioning in the Microsoft land is truly a farce.

**Comment by boris-kolpackov** [comment]
Related to this is specifying the MSVC toolchain for Clang to use. Some excerpts from a recent Slack chat by @kamrann:

> Did a bit of debugging of clang, there are `/vctoolsdir` and `/vctoolsversion` arguments but they're specific to clang-cl. Simplest solution that achieves the same thing with the regular driver is to add `VCToolsInstallDir=path/to/toolchain` to the environment. For whatever reason, it doesn't look like it can be controlled via regular driver command line.

Also note that `-stdlib++-isystem<directory>` doesn't appear to do the correct thing while running from the correctly setup VS Command Prompt works as expected. The Slack thread is here: https://cpplang.slack.com/archives/CDJ0Z991S/p1687516931916039
